sdk/core/src/services/AffinidiVaultStorageService.ts

```diff
@@ -0,0 +1,184 @@
+  }
+
+  private async _authorizeVcVault(): Promise<string> {
+    const { body } = await this._vaultApiService.createDidAuthRequest({ audienceDid: this._audienceDid })
+    const responseToken = await this._didAuthService.createDidAuthResponseToken(body)
+    return responseToken
+  }
+
+  private async _encryptCredentials(credentials: any[]): Promise<VaultCredential[]> {
+    const publicKeyBuffer = this._keysService.getOwnPublicKey()
+    const privateKeyBuffer = this._keysService.getOwnPrivateKey()
+    const encryptedCredentials: VaultCredential[] = []
+
+    for (const credential of credentials) {
+      let credentialId = credential?.id
+      if (!isW3cCredential(credential)) {
+        credentialId = credential?.data?.id
+      }
+
+      const credentialIdHash = await this._platformEncryptionTools.computePersonalHash(privateKeyBuffer, credentialId)
+
+      const typeHashes = []
+      if (isW3cCredential(credential)) {
+        for (const credentialType of credential.type) {
+          const typeHash = await this._platformEncryptionTools.computePersonalHash(privateKeyBuffer, credentialType)
+          typeHashes.push(typeHash)
+        }
+      }
+
+      const cyphertext = await this._platformEncryptionTools.encryptByPublicKey(publicKeyBuffer, credential)
+
+      encryptedCredentials.push({
+        credentialId: credentialIdHash,
+        credentialTypes: typeHashes,
+        payload: cyphertext,
+      })
+    }
+
+    return encryptedCredentials
+  }
+
+  private async _computeTypesHashes(types: string[][]): Promise<string[][]> {
+    const privateKeyBuffer = this._keysService.getOwnPrivateKey()
+
+    const hashedTypes: string[][] = []
+    for (const subset of types) {
+      const hashedSubset: string[] = []
+      for (const type of subset) {
+        const hashedType = await this._platformEncryptionTools.computePersonalHash(privateKeyBuffer, type)
+        hashedSubset.push(hashedType)
+      }
+
+      hashedTypes.push(hashedSubset)
+    }
+
+    return hashedTypes
+  }
+
+  private async _decryptCredentials(encryptedCredentials: VaultCredential[]): Promise<any[]> {
+    const credentials: any[] = []
+
+    for (const encryptedCredential of encryptedCredentials) {
+      const credential = await this._decryptCredential(encryptedCredential)
+      credentials.push(credential)
+    }
+
+    return credentials
+  }
+
+  private async _decryptCredential(encryptedCredential: VaultCredential): Promise<any> {
+    const privateKeyBuffer = this._keysService.getOwnPrivateKey()
+
+    const credential = await this._platformEncryptionTools.decryptByPrivateKey(
+      privateKeyBuffer,
+      encryptedCredential.payload,
+    )
+
+    return credential
+  }
+
+  public async saveCredentials(credentials: any[], storageRegion: string): Promise<VaultCredential[]> {
+    const token = await this._authorizeVcVault()
+
+    const responses = []
+
+    const encryptedCredentials = await this._encryptCredentials(credentials)
```

Comment:
the did auth has some expiration time set by the server if I remember correctly. for this exact reason, I remember setting up token expiration checking / re-auth logic in some frontend. please see: https://gitlab.com/affinidi/foundational/affinity-verifier-portal/-/blob/main/src/services/MessageService.ts#L53

Here after the did auth we are doing N encryption and HTTP requests to the backend with the same token. According to the clients computing power, network speed that expiration can happen. Causing some operations to fail. For that reason; I think it would be better to implement something like the link I've sent, and calling the did-auth everytime before making a request to the server.

Reply:
@Yengas, the same logic was used in bloom vault.

The expiration time is AFAIK not measured in seconds, and if this method takes so long to complete that token expires, we'll probably have some other problem.

Reply:
That actually reasonable as our sdk allow to store thousands and even more credentials at same time and we cannot just reject this. What we can do that combine token refresh logic with idea  described https://github.com/affinityproject/affinidi-core-sdk/pull/117#discussion_r674595396 and then we actually can reuse token for same `audienceDid` requests in short period of time by keeping it in memory and refresh when it got expired.

Reply:
But that is going to be as enhancement as for me. Just don't want increase size of current PR

Reply:
Sure I agree it is not exactly related to this PR. But it is maybe a blocker to release v6(I need to check expiration date).

I created [FTL-1770](https://replika.atlassian.net/browse/FTL-1770), I will try to find expiration date and add more info do that task regarding optimizations. If it is a blocker, we should do it before removing the beta suffix. Thanks guys.